tests/integration/test_single.py

```diff
@@ -43,187 +45,25 @@
-                            "-QT",
-                            "-fverilog",
-                            file_stem + ".sv",
-                            "-pstat",
-                        ]
-                    ),
-                    shell=True,
-                    text=True,
-                    stdout=subprocess.PIPE,
-                    stderr=subprocess.PIPE,
-                ) as syn_process:
-                    syn_process.wait()
-                    stdout = syn_process.stdout.read()
-                    stderr = syn_process.stderr.read()
-
-                self.assertFalse(stderr.strip())
-                stats = stdout[stdout.find("Printing statistics.") :]
-
-                def snake_case(text):
-                    return re.sub(r"[\W_]+", "_", text).strip("_").lower()
-
-                lines = stats.strip().splitlines()
-                data = {}
-
-                for line in lines:
-                    if ":" in line:
-                        key, value = line.split(":")
-                        key = snake_case(key).split("number_of_")[-1]
-                        value = int(value.strip())
-                    else:
-                        try:
-                            index = line.find("$") + 10
-                            value = int(line[index:].strip())
-                            key = line[:index].strip()[1:]
-                            data[key] = value
-
-                        except ValueError:
-                            continue
-
-                    data[key] = value
-                for key, value in data.items():
-                    statistics[key] = value
-
-            TestComplete.statistics.append(statistics)
-
-    @classmethod
-    def tearDownClass(cls):
-        cls.statistics.sort(key=lambda e: e["Test Name"])
-        df = pd.DataFrame(cls.statistics, columns=cls.statistics[0].keys())
-        title = f" Statistics for {cls.__class__.__name__} "
-        table = df.to_markdown()
-        table_width = len(table.partition("\n")[0])
-        pad = table_width - len(title)
-        result = (
-            "\n" + "=" * (pad // 2) + title + "=" * (pad // 2 + pad % 2) + "\n" + table
-        )
-        logging.warning(result)
-        if cls.write:
-            stats_file_name = (
-                os.path.commonprefix(
-                    list(map(lambda e: e["Test Name"], cls.statistics))
-                ).replace("::", "")
-                + ".csv"
-            )
-            with open(Path(__file__).parent / stats_file_name, mode="w") as f:
-                f.write(df.to_csv(index=False))
+        BaseTest.multi_perf(self, funcs, test_cases)
 
     @parameterized.expand(
         input=PARAMETERS,
         name_func=name_func,
     )
-    def test_single_correctness(
-        self, func: FunctionType, test_cases: list[Union[tuple[int, ...], int]]
+    def test_multi_correct(
+        self, funcs: list[FunctionType], test_cases: list[Union[tuple[int, ...], int]]
```

Comment:
> `list[Union[tuple[int, ...], int]]`

Is there anyway to simplify this data structure? i.e. with a testcase class?

Perhaps you can also add functions to it so that the user can add more specific test-cases that are then translated using p2v

Reply:
I could see a case where I should typedef this type. However, I'm curious as to what kind of member functions a testcase class would have. Could you give some examples?

Reply:
What is the reasoning behind the `Union`?

My thought process is you can pass the Python function you're testing and a set of inputs. You can then call the testing function to generate a set of outputs to compare it with the module in question.

Reply:
> What is the reasoning behind the `Union`?

So that if a function takes 1 argument, I can pass `10` or `(10,)`, instead of just the latter.

> My thought process is you can pass the Python function you're testing and a set of inputs. You can then call the testing function to generate a set of outputs to compare it with the module in question.

Yes this is what I do, as seen https://github.com/WorldofKerry/Python2Verilog/blob/6e63f7c7d66bbe3dff08d30e301395d96cf2bd45/tests/integration/test_single.py#L16-L26

Reply:
I see, makes sense.